prov/usnic/src/usdf_fabric.c

```diff
@@ -341,6 +341,94 @@ static int validate_modebits(uint32_t version, const struct fi_info *hints,
 	return FI_SUCCESS;
 }
 
+static int usdf_alloc_fid_nic(struct fi_info *fi,
+			struct usd_device_attrs *dap)
+{
+	int ret;
+	struct fid_nic *nic = NULL;
+	struct fi_device_attr *da = NULL;
+	struct fi_bus_attr *ba = NULL;
+	struct fi_link_attr *la = NULL;
+
+	nic = calloc(1, sizeof(*nic));
+	if (!nic)
+		goto nomem;
+	nic->fid.fclass = FI_CLASS_NIC;
+
+	da = calloc(1, sizeof(*da));
+	if (!da)
+		goto nomem;
+	nic->device_attr = da;
+	da->name = strdup(dap->uda_devname);
+	ret = asprintf(&da->device_id, "%s (%s)",
+		usd_devid_to_pid(dap->uda_vendor_id,
+				dap->uda_device_id),
+		usd_devid_to_nicname(dap->uda_vendor_id,
+				dap->uda_device_id));
+	if (ret < 0) {
+		USDF_DBG(
+			"asprintf failed while creating device_id\n");
+		ret = -ENOMEM;
```

Comment:
I think you're missing a `goto` in this cleanup path. You set `ret` but then overwrite it immediately afterward.